pyvista/utilities/reader.py

```diff
@@ -0,0 +1,368 @@
+"""Fine-grained control of reading data files."""
+
+from typing import Any
+
+from pyvista.utilities import wrap, get_ext
+from pyvista import _vtk
+
+
+def get_reader(filename):
+    """Get a reader for fine-grained control of reading data files.
+
+    Supported file types and Readers:
+    * .vti: :class:`pyvista.XMLImageDataReader`
+    * .pvti: :class:`pyvista.XMLPImageDataReader`
+    * .vtr: :class:`pyvista.XMLRectilinearGridReader`
+    * .pvtr: :class:`pyvista.XMLPRectilinearGridReader`
+    * .vtu: :class:`pyvista.XMLUnstructuredGridReader`
+    * .pvtu: :class:`pyvista.XMLPUnstructuredGridReader`
+    * .vtp: :class:`pyvista.XMLPolyDataReader`
+    * .vts: :class:`pyvista.XMLStructuredGridReader`
+    * .vtm: :class:`pyvista.XMLMultiBlockDataReader`
+    * .vtmb: :class:`pyvista.XMLMultiBlockDataReader`
+    * .case: :class:`pyvista.EnSightReader`
+    * .foam: :class:`pyvista.OpenFOAMReader`
+    * .ply: :class:`pyvista.PLYReader`
+    * .obj: :class:`pyvista.OBJReader`
+    * .stl: :class:`pyvista.STLReader`
+    * .vtk: :class:`pyvista.VTKDataSetReader`
+    * .pvtk: :class:`pyvista.VTKPDataSetReader`
+    * .g: :class:`pyvista.BYUReader`
+    * .facet: :class:`pyvista.FacetReader`
+    * .p3d: :class:`pyvista.Plot3DMetaReader`
+    * .tri: :class:`pyvista.BinaryMarchingCubesReader`
+
+    Parameters
+    ----------
+    filename : str
+        The string path to the file to read.
+
+    Returns
+    -------
+    :class:`pyvista.BaseReader`
+        A subclass of `pyvista.BaseReader` is returned based on file type.
+
+    """
+    ext = get_ext(filename)
+
+    try:
+        Reader = CLASS_READERS[ext]
+    except KeyError:
+        raise ValueError(f"{ext} not supported")
+
+    return Reader(filename)
+
+class BaseReader:
+    """The base reader class."""
+
+    _class_reader: Any = None
+
+    def __init__(self, filename):
+        """Initialize Reader by setting filename."""
+        self._reader = self._class_reader()
+        self.filename = filename
+        self._set_filename(filename)
+
+    @property
+    def reader(self):
+        """Return the vtk Reader object."""
+        if self._reader is None:
+            raise NotImplementedError
+        return self._reader
+
+    def _set_filename(self, filename):
+        """Set filename and update reader."""
+        # Private method since changing file type requires a
+        # different subclass.
+        self.filename = filename
+        self.reader.SetFileName(filename)
+        self.update()
```

Comment:
I found that the OpenFOAM reader requires this to call `reader.reader.UpdateInformation()` to be able to set the time points correctly.  I need to test that this works for all the other readers before changing.

Reply:
The example from #1511 was really helpful here.